docs/adr/adr-009-telemetry.md

```diff
@@ -0,0 +1,466 @@
+        attribute.Int("axis", q.source),
+        attribute.Int("x", q.x),
+        attribute.Int("y", q.y),
+        attribute.Int("size", len(q.roots)),
+    ))
+```
+
+> The above is only examples related to our code and is a subject to change.
+
+Here is the result of the above code sending traces visualized on Jaeger UI
+![tracing](img/trace-jaeger.png)
+
+#### Backends connection
+
+Example for Jaeger
+
+```go
+  // Create the Jaeger exporter
+  exp, err := jaeger.New(jaeger.WithCollectorEndpoint(jaeger.WithEndpoint(url)))
+  if err != nil {
+      return nil, err
+  }
+ // then the tracer provider
+    tp := tracesdk.NewTracerProvider(
+        // Always be sure to batch in production.
+        tracesdk.WithBatcher(exp),
+        // Record information about this application in a Resource.
+        tracesdk.WithResource(resource.NewWithAttributes(
+            semconv.SchemaURL,
+            semconv.ServiceNameKey.String(service),
+            attribute.String("environment", environment),
+            attribute.Int64("ID", id),
+        )),
+    )
+ // and set it globally to be used across packages
+    otel.SetTracerProvider(tp)
+ 
+ // then close it elsewhere
+    tp.Shutdown(ctx)
+```
+
+We decided to use OTLP backend, and it is almost similar in terms of setup.
+
+### Metrics Design
+
+Metrics allows collecting time-series data from different measurable points in the application. Every measurable can
+be covered via 6 instruments OpenTelemetry provides:
+
+* ___Counter___ - synchronous instrument that measures additive non-decreasing values.
+* ___UpDownCounter___ - synchronous instrument which measures additive values that increase or decrease with time.
+* ___Histogram___ - synchronous instrument that produces a histogram from recorded values.
+* ___CounterObserver___ - asynchronous instrument that measures additive non-decreasing values.
+* ___UpDownCounterObserver___ - asynchronous instrument that measures additive values that can increase or decrease with time.
+* ___GaugeObserver___ - asynchronous instrument that measures non-additive values for which sum does not produce a meaningful correct result.
+
+#### Metrics Integration Example
+
+Consider we want to report the current network height as a metric.
+
+First of all, the global pkg meter has to be defined in the code related to the desired metric. In our case it is `header` pkg.
+
+```go
+var meter = global.MeterProvider().Meter("header")
+```
+
+Next, we should understand what instrument to use. On the first glance, for chain height a ___Counter___ instrument should
+fit, as it is a non-decreasing value, and then we should think whether we need a sync or async version of it. For our case,
+both would work and its more the question of precision we want. Sync metering would report every height change, while
+the async would poke `header` pkg API periodically to get the metered data. For our example, we will go with the latter.
```

Comment:
This section is missing the point that Counter will be iterated from multiple sources, that might include same application being restarted. For specified case only async Gauge can be used to reflect current network height.